### PR TITLE
chore: migrate from yarn to pnpm

### LINKED ISSUE #380: Add support of `pnpm` package managers
Upgrade and migrate currently support `yarn` and `npm` project, but there are `pnpm` ones.

The difference is:
- `npm install <package>`, `npm install -g <package>`
- `pnpm add <package>`, `pnpm add -g <package>`
- `yarn add <package>`, `yarn global add <package>`

Could be useful:
- https://github.com/egoist/detect-package-manager

### PR DESCRIPTION
**Pull request by lynchbomb**
Closes #380 
This PR migrates us from yarn1 to pnpm workspace with lerna

### CODE REVIEW

File: .github/dependabot.yml [hunk 1]
```diff
@@ -1,25 +1,25 @@
...
-    versions:
-    - 7.19.0
\ No newline at end of file
+  - package-ecosystem: npm
+    directory: '/'
+    schedule:
+      interval: monthly
+    open-pull-requests-limit: 20
+    versioning-strategy: increase-if-necessary
+    ignore:
```
Comment (egorio) [inline .github/dependabot.yml:9]: I think you can remove `ignore`

File: package.json [hunk 2]
```diff
@@ -32,18 +26,54 @@
...
+    "lerna": "^5.5.2",
     "pre-commit": "^1.2.2",
     "prettier": "^2.7.1",
     "typescript": "4.7.4",
-    "vitest": "^0.22.1"
+    "vitest": "^0.23.4",
+    "@types/commander": "^2.12.2",
+    "@types/lodash.get": "^4.4.7",
+    "@types/sarif": "^2.1.4",
+    "lodash.get": "^4.4.2",
```
Comment (egorio) [inline package.json:37]: This list is usually sorted you you use `add`.

File: packages/cli/package.json [hunk 3]
```diff
@@ -35,12 +35,13 @@
   },
   "dependencies": {
     "@microsoft/api-documenter": "^7.13.68",
-    "@rehearsal/migrate": "^0.0.36",
-    "@rehearsal/reporter": "^0.0.36",
-    "@rehearsal/upgrade": "^0.0.36",
+    "@rehearsal/migrate": "workspace:*",
+    "@rehearsal/reporter": "workspace:*",
+    "@rehearsal/upgrade": "workspace:*",
+    "@rehearsal/migration-graph": "workspace:*",
```
Comment (egorio) [inline packages/cli/package.json:41]: 👍

File: packages/cli/src/utils.ts [hunk 4]
```diff
@@ -149,63 +149,125 @@ export async function isYarnManager(): Promise<boolean> {
...
+export async function getModuleManager(): Promise<'yarn' | 'npm' | 'pnpm'> {
   const isYarn = await isYarnManager();
-  // check if npm or yarn
-  const binAndArgs = {
-    bin: isYarn ? 'yarn' : 'npm',
-    args: isYarn
-      ? ['add', '-D', `${devDep}`, '--ignore-scripts']
-      : ['install', `${devDep}`, '--save-dev', '--ignore-scripts'],
-  };
+  const isPnpm = await isPnpmManager();
```
Comment (egorio) [inline packages/cli/src/utils.ts:190]: You can use Promise.all like here:
https://github.com/rehearsal-js/action/blob/master/src/pm.ts

File: package.json (hunk 2 above)
Comment (lynchbomb) [reply package.json:37]: yeah merge conflicts. im just going to add an auto sort for package.json after this PR

File: packages/cli/src/utils.ts [hunk 5]
```diff
@@ -149,63 +149,125 @@ export async function isYarnManager(): Promise<boolean> {
...
+
+  if (isPnpm) {
+    return 'pnpm';
+  }
+
+  return 'npm';
+}
 
-  await execa(binAndArgs.bin, binAndArgs.args);
+export async function getModuleManagerInstaller(
```
Comment (egorio) [inline packages/cli/src/utils.ts:203]: To `@rehearsal/utils`. I will be able to use it in GitHub Action :)

File: packages/plugins/package.json [hunk 6]
```diff
@@ -25,21 +26,21 @@
...
+    "@rehearsal/service": "workspace:*",
+    "@rehearsal/utils": "workspace:*",
+    "@rehearsal/reporter": "workspace:*",
     "eslint": "8.22.0"
   },
   "devDependencies": {
-    "@rehearsal/reporter": "^0.0.36",
-    "@rehearsal/utils": "^0.0.36",
-    "@types/eslint": "^8.4.1",
+    "vitest": "^0.23.4",
```
Comment (egorio) [inline packages/plugins/package.json:41]: `workspace:*` ?

File: packages/reporter/src/reporter.ts [hunk 7]
```diff
@@ -11,7 +11,7 @@ import { type ProcessedFile, type Report, type ReportFormatter, type ReportItem
 export class Reporter {
   readonly basePath: string;
 
-  private report: Report;
+  public report: Report;
```
Comment (egorio) [inline packages/reporter/src/reporter.ts:14]: Don't do this public just for test :(

File: packages/upgrade/test/upgrade.test.ts [hunk 8]
```diff
@@ -35,7 +35,7 @@ describe('Test upgrade', async function () {
   });
 
   test('should output the correct data from upgrade', () => {
-    const report = (reporter as any).report as Report;
+    const { report } = reporter;
```
Comment (egorio) [inline packages/upgrade/test/upgrade.test.ts:38]: Please keep the report private, you can use `/* eslint-disable ... */` here

File: packages/reporter/src/reporter.ts (hunk 7 above)
Comment (lynchbomb) [reply packages/reporter/src/reporter.ts:14]: yeah i was going back and forth on this. I think its worse doing the type cast in tests. it really doesnt matter that this is public

File: packages/upgrade/test/upgrade.test.ts (hunk 8 above)
Comment (lynchbomb) [reply packages/upgrade/test/upgrade.test.ts:38]: see prev. comment. doing the type cast is worse. why does it matter if the report is public?

File: packages/cli/src/utils.ts (hunk 4 above)
Comment (lynchbomb) [reply packages/cli/src/utils.ts:190]: oh nice. ok

**Review by egorio — APPROVED** [review]

**Pull request merged**